Update to 1.13

buchwasa opened the pull request:
## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
Huge thanks to NetherGames and Nukkit ~~Steadfast2~~ for providing most of the information needed to get a basic implementation done. This PR is needed to update PM to 1.13. Most things that are needed to update PM to 1.13 have been implemented to a working state. Quite honestly, I wouldn't say this is 100% safe to merge yet, I haven't had sufficient testing.

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
- NetworkBinaryStream has 4 new methods: getSkin() : SkinData, putSkin(SkinData), getSkinImage(), and putSkinImage(SkinImage)
- Player->sendRespawnPacket accepts the respawnState parameter
- Added the `SkinImage` and `SkinAnimation` classes
- Added the SkinAdapter interface to deal with the data received.
- Added the LegacySkinAdapter to convert all incoming skins to legacy (for BC reasons)
- Added the SkinAdapterSingleton class with get : SkinAdapter and set(SkinAdapter) allowing developers to create their own skin adapters
- PlayerListEntry no longer accepts a Skin and accepts only SkinData, as well as PlayerSkinPacket

### Behavioural changes
<!-- Any change in how the server behaves, or its performance? -->
It's still a server, no worries
## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->

## Follow-up
- Vanilla resources will need updated

## Tests
<!--
Details should be provided of tests done. Simply saying "tested" or equivalent is not acceptable.

Attach scripts or actions to test this pull request, as well as the result
-->
This hasn't be thoroughly tested, I've just done some here and there testing and it seems to work okay.

Commit:
Initial update to 1.13

Commit:
Moved runtime_block_states to vanilla

Commit:
Fixed changing skins crashing the server

Commit:
Clean up duplicated skin entries

Commit:
Fixed persona skins being messed up on join

Commit:
Removed some useless casts

Commit:
Start saving new skin data

Commit:
Added CompletedUsingItemPacket

Commit:
Fixed stupid mistake

Commit:
Fixed items, but I'm not so sure how well this'll work out, it'll need improved in the future

Commit:
Fixed respawning

Commit:
Fixed crafting

Commit:
Updated ResourcePackType

Commit:
Merge branch 'stable' of https://github.com/pmmp/PocketMine-MP into stable

Commit:
created method sendCompletedUsingItemPacket for less duplication

Commit:
Cape data can be null

Commit:
Fixed FloatingTextParticles

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -2638,38 +2668,14 @@ public function handleInventoryTransaction(InventoryTransactionPacket $packet) :
-
-								return true;
-							}
-
```
Why remove this whole block of code, and yet leave behind the case?

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -834,12 +839,10 @@ public function getLocale() : string{
 	 * @return bool
 	 */
-	public function changeSkin(Skin $skin, string $newSkinName, string $oldSkinName) : bool{
+	public function changeSkin(Skin $skin) : bool{
```
These are here for future usage. If their removal isn't related to a protocol change, leave them alone.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -1913,12 +1925,22 @@ public function handleLogin(LoginPacket $packet) : bool{
+			$packet->clientData["PremiumSkin"] ?? false,
+			$packet->clientData["PersonaSkin"] ?? false,
+			$packet->clientData["CapeOnClassicSkin"] ?? false,
+			$packet->clientData["CapeId"] ?? ""
```
RIP backwards compatibility.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -285,8 +290,8 @@ public static function isValidUserName(?string $name) : bool{
-	/** @var PlayerCursorInventory */
-	protected $cursorInventory;
+	/** @var PlayerUIInventory */
+	protected $playerUIInventory;
```
BC-breaking change.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -2520,7 +2542,15 @@ public function handleInventoryTransaction(InventoryTransactionPacket $packet) :
 						}
 
-						$this->setUsingItem(true);
+						if($this->startAction === -1){
```
`isUsingItem()` should be used here.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -3804,16 +3819,19 @@ protected function addDefaultWindows(){
+		$this->playerUIInventory = new PlayerUIInventory($this);
+		$this->addWindow($this->playerUIInventory, ContainerIds::UI, true);
 
+		$this->craftingGrid = new CraftingGrid($this->playerUIInventory, CraftingGrid::SIZE_SMALL);
```
BC-breaking change.

dktapps commented on src/pocketmine/entity/Human.php:
```diff
@@ -117,10 +120,16 @@ public function __construct(Level $level, CompoundTag $nbt){
-			$skinTag->getString("GeometryName", ""),
-			$skinTag->getByteArray("GeometryData", "")
+			$skinTag->getString("SkinResourcePatch", ""),
+			new SerializedImage($skinTag->getInt("SkinImageHeight"), $skinTag->getInt("SkinImageWidth"), $skinTag->getByteArray("Data")),
```
`SkinImageHeight`, `SkinImageWidth` won't exist in worlds loaded from 1.12. This will cause a crash.

There is a backwards compatibility patch for `Data` that has been removed here that will also cause crashes on old data.

dktapps commented on src/pocketmine/entity/Human.php:
```diff
@@ -117,10 +120,16 @@ public function __construct(Level $level, CompoundTag $nbt){
+			$skinTag->getString("SkinResourcePatch", ""),
+			new SerializedImage($skinTag->getInt("SkinImageHeight"), $skinTag->getInt("SkinImageWidth"), $skinTag->getByteArray("Data")),
+			"", //TODO: animations
+			new SerializedImage($skinTag->getInt("CapeImageHeight"), $skinTag->getInt("CapeImageWidth"), $skinTag->getByteArray("CapeData")),
```
`CapeImageHeight`, `CapeImageWidth` won't exist in worlds loaded from 1.12. This will cause a crash.

dktapps commented on src/pocketmine/entity/Human.php:
```diff
@@ -117,10 +120,16 @@ public function __construct(Level $level, CompoundTag $nbt){
+			new SerializedImage($skinTag->getInt("CapeImageHeight"), $skinTag->getInt("CapeImageWidth"), $skinTag->getByteArray("CapeData")),
+			$skinTag->getByteArray("GeometryData", ""),
+			$skinTag->getByteArray("AnimationData", ""),
+			$skinTag->getByte("PremiumSkin") === 1,
```
This won't exist on 1.12 world data.

dktapps commented on src/pocketmine/entity/Human.php:
```diff
@@ -117,10 +120,16 @@ public function __construct(Level $level, CompoundTag $nbt){
+			$skinTag->getByteArray("GeometryData", ""),
+			$skinTag->getByteArray("AnimationData", ""),
+			$skinTag->getByte("PremiumSkin") === 1,
+			$skinTag->getByte("PersonaSkin") === 1,
```
nor this

dktapps commented on src/pocketmine/entity/Human.php:
```diff
@@ -117,10 +120,16 @@ public function __construct(Level $level, CompoundTag $nbt){
+			$skinTag->getByteArray("AnimationData", ""),
+			$skinTag->getByte("PremiumSkin") === 1,
+			$skinTag->getByte("PersonaSkin") === 1,
+			$skinTag->getByte("CapeOnClassic") === 1,
```
nor this

dktapps commented on src/pocketmine/entity/Skin.php:
```diff
@@ -24,6 +24,8 @@
 namespace pocketmine\entity;
```
BC in this file is completely trashed. I don't know where to start. It doesn't look like any attempt has been made to preserve BC whatsoever.

dktapps commented on src/pocketmine/inventory/BaseUIInventory.php:
```diff
@@ -0,0 +1,116 @@
+use pocketmine\item\ItemFactory;
+use pocketmine\Player;
+
+class BaseUIInventory extends BaseInventory{
```
The introduction of this class pushes network internals off onto plugin developers for the sake of catering for one change: the addressing of fake inventories is now mapped to window ID `124` and using magic slot numbers, instead of negative window IDs and normal slot numbers.

This is an overcomplicated solution for such a problem which would be better solved at a network-local level with a more isolated hack that doesn't touch the API.

dktapps commented on src/pocketmine/inventory/CraftingGrid.php:
```diff
@@ -29,134 +29,23 @@
-	public const SIZE_SMALL = 2;
-	public const SIZE_BIG = 3;
+class CraftingGrid extends BaseUIInventory{
+	public const SIZE_SMALL = 4;
```
This is a BC-break.

dktapps commented on src/pocketmine/inventory/CraftingGrid.php:
```diff
@@ -29,134 +29,23 @@
-	public const SIZE_BIG = 3;
+class CraftingGrid extends BaseUIInventory{
+	public const SIZE_SMALL = 4;
+	public const SIZE_BIG = 9;
```
and this

dktapps commented on src/pocketmine/inventory/CraftingGrid.php:
```diff
@@ -29,134 +29,23 @@
-	 *
-	 * @return Item
-	 */
-	public function getIngredient(int $x, int $y) : Item{
```
BC break here

dktapps commented on src/pocketmine/inventory/CraftingGrid.php:
```diff
@@ -29,134 +29,23 @@
-	 *
-	 * @return int
-	 */
-	public function getRecipeWidth() : int{
```
BC break here

dktapps commented on src/pocketmine/inventory/CraftingGrid.php:
```diff
@@ -29,134 +29,23 @@
-	 * Returns the height of the recipe we're trying to craft, based on items currently in the grid.
-	 * @return int
-	 */
-	public function getRecipeHeight() : int{
```
BC break here.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/RuntimeBlockMapping.php:
```diff
@@ -50,20 +53,27 @@ private function __construct(){
+			/** @var CompoundTag $tag */
+			$tag = (new BigEndianNBTStream())->read(file_get_contents(\pocketmine\RESOURCE_PATH . "vanilla/runtime_block_states.dat"));
+		}catch(BinaryDataException $e){
+			throw new \RuntimeException("", 0, $e);
```
what exactly is the point of this?

dktapps commented on src/pocketmine/utils/SkinAnimation.php:
```diff
@@ -0,0 +1,52 @@
+	/** @var SerializedImage */
+	private $image;
+	/** @var int */
+	private $type;
```
This is undocumented and has no constants.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/ResourcePackType.php:
```diff
@@ -30,9 +30,13 @@ private function __construct(){
+	public const RESOURCES = 6;
+	public const SKINS = 7;
+	public const WORLD_TEMPLATE = 8;
+	public const COUNT = 9;
```
Cargo-cult code.

It's common practice in C++ to add a dummy member to the end of an enum for range-checking, like so:

```
enum things: uint8_t {
    thing1,
    thing2,
    thing3,
    number_of_things
}
```
Here, this doesn't make any sense, because this is not an enum.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/ResourcePackType.php:
```diff
@@ -30,9 +30,13 @@ private function __construct(){
+	public const ADDON = 1;
+	public const CACHED = 2;
+	public const COPY_PROTECTED = 3;
+	public const BEHAVIOR = 4;
```
BC-break here. This should be `BEHAVIORS`.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/PlayerListEntry.php:
```diff
@@ -34,12 +34,18 @@ class PlayerListEntry{
+	/** @var int */
+	public $buildPlatform = 0; //Unknown
+	/** @var Skin */
+	public $skin;
```
This change is utterly redundant.

dktapps commented on src/pocketmine/network/mcpe/protocol/ProtocolInfo.php:
```diff
@@ -185,5 +185,13 @@ interface ProtocolInfo{
+	public const ANVIL_DAMAGE_PACKET = 0x8d;
+	public const COMPLETED_USING_ITEM_PACKET = 0x8e;
+	public const NETWORK_SETTINGS_PACKET = 0x8f;
+	public const PLAYER_AUTH_INPUT_PACKET = 0x90;
```
These packets seem to be missing.

dktapps commented on src/pocketmine/network/mcpe/protocol/PlayerListPacket.php:
```diff
@@ -29,6 +29,8 @@
 use pocketmine\network\mcpe\NetworkSession;
 use pocketmine\network\mcpe\protocol\types\PlayerListEntry;
+use pocketmine\utils\SerializedImage;
+use pocketmine\utils\SkinAnimation;
```
These imports appear to be redundant.

dktapps commented on src/pocketmine/network/mcpe/protocol/PacketPool.php:
```diff
@@ -23,6 +23,7 @@
 
 namespace pocketmine\network\mcpe\protocol;
 
+use pocketmine\network\mcpe\protocol\CompletedUsingItemPacket;
```
This import is redundant.

dktapps commented on src/pocketmine/network/mcpe/protocol/CompletedUsingItemPacket.php:
```diff
@@ -0,0 +1,66 @@
+	/** @var int */
+	public $action;
+
+	public function decodePayload(){
```
Please apply `void` typehints here for easier integration with `master`.

dktapps commented on src/pocketmine/network/mcpe/protocol/CompletedUsingItemPacket.php:
```diff
@@ -0,0 +1,66 @@
+		$this->action = $this->getLInt();
+	}
+
+	public function encodePayload(){
```
Please apply `void` typehints here for easier integration with `master`.

dktapps commented on src/pocketmine/network/mcpe/protocol/CompletedUsingItemPacket.php:
```diff
@@ -0,0 +1,66 @@
+	/** @var int */
+	public $itemId;
+	/** @var int */
+	public $action;
```
Recent style changes prefer private fields and public getters & a `::create()` function, e.g. https://github.com/pmmp/PocketMine-MP/blob/cac3c356a5a83cf6cada0269e86f8d8d11e80c52/src/pocketmine/network/mcpe/protocol/ClientCacheStatusPacket.php#L33-L47

dktapps commented on src/pocketmine/network/mcpe/NetworkSession.php:
```diff
@@ -23,6 +23,7 @@
 
 namespace pocketmine\network\mcpe;
 
+use pocketmine\network\mcpe\protocol\CompletedUsingItemPacket;
```
Imports should be sorted lexicographically ascending.

dktapps commented on src/pocketmine/item/Consumable.php:
```diff
@@ -26,6 +26,7 @@
 use pocketmine\block\Block;
 use pocketmine\entity\EffectInstance;
 use pocketmine\entity\Living;
+use pocketmine\Player;
```
This change is superfluous.

dktapps reviewed (changes requested):
I haven't given this a full review, but I've seen enough to haze it so far.

I want to say that this is a worthwhile first attempt at an update. However, several things have significantly overblown complexity and this PR makes no effort whatsoever to preserve backwards compatibility in any arena.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -2520,7 +2542,15 @@ public function handleInventoryTransaction(InventoryTransactionPacket $packet) :
+						}
+
+						$this->setUsingItem(false);
+						if($item->onUse($this)){
```
This change creates a behavioural break for food consumption. The events and benefits are supposed to apply after the item is finished being eaten (during the CONSUME transaction type), not when the player just started using the item.

yoraze commented:
@dktapps, I don’t think he will understand your comments on this PR. As I understand it, this is a "very good copy+paste from NetherGames software".

dktapps commented:
Thanks, there's no need for that. I think if that's the case it will make itself apparent pretty quickly without toxicity.

Driesboy commented:
If we wouldn't like our code being used we would have made it opensource. Simple it is.

SOF3 commented:
Please stop spamming this issue with irrelevant flame war. If you have unsettled personal arguments, please resolve them in private.

DaPigGuy commented:
Don't forget to update the AvailableCommandsPacket constants :)

SleepSpace9 commented:
ItemFrame blocks should also not be forgotten ("Can now be placed on the top and bottom of blocks"). The new metadata for their sides needs to be updated, otherwhise they behave strangely when placed and idk if it will save wrong nbt data to the world too.

yoraze commented:
The author of this PR copied a broken block palette from SteadFast2, because of this, possibly, similar problems arise.

Commit:
Removed root directory runtime states

Commit:
Clean up unused imports

Commit:
Fixed convertToLegacyName not using variables,  add typehints to it too

Commit:
Add typehints to completedusingitempacket

Commit:
Moved skin back

Not too sure what I planned on doing with that

Commit:
Removed count const and fixed BC break

Commit:
No need for a try catch statement around reading a file

Commit:
Switch to isUsingItem()

I don't plan on using this, as you informed me on discord that I didn't need the extra step, and it's a BC break anyway

Commit:
Simplify if null statement

Commit:
Removed PlayerUIInventory and BaseUIInventory

Wumpotamus commented:
The CraftingGrid API needs more testing, but PlayerCursorInventory should work just fine, I was able to test that. I'm working on removing the getCompletedAction() stuff, since it was a bad implementation and isn't actually needed to work at all.

Commit:
Clean up whitespace disaster and add getIngredient back

Encritary commented:
@Wumpotamus On my test server, I check if item is consumable and if it is, I do consume logic and set action to ``ACTION_CONSUME``, otherwise the action is ``ACTION_UNKNOWN``. No changes in ``pocketmine\item`` at all.

dktapps commented:
CompleteUsingItemPacket is superfluous.

Encritary commented:
@dktapps is it?

dktapps commented:
It's useless. According to my inside sources, it's only used for statistics.

The bug with food is because the CONSUME type for ReleaseItemTransaction doesn't exist anymore. It's now merged with the regular RELEASE type. The changes needed to address that are significantly less than seen here.

dktapps committed:
CompletedUsingItemPacket is superfluous

strich3 commented:
when is the 1.13 update Best regards Potato.net Network! 😉

dktapps committed:
cleanup eating clusterfuck

dktapps committed:
Player: remove unused import

dktapps committed:
reverse BC-breaking change to Player->changeSkin()
this change would cause LSP violations for subclasses

dktapps committed:
reverse unnecessary formatting changes

dktapps committed:
superfluous import

Ifera committed:
Merge pull request #3 from dktapps/drew-1.13

clean up 1.13 a little

yoraze commented on src/pocketmine/network/mcpe/protocol/types/PlayerListEntry.php:
```diff
@@ -40,6 +40,12 @@ class PlayerListEntry{
 	/** @var string */
 	public $platformChatId = "";
+	/** @var int */
+	public $buildPlatform = 0; //Unknown
```
Could you rename this to `deviceOs`? I think this is a more suitable name for this field.
And it's better to use -1 as the default value: -1 = UNKNOWN_OS

yoraze commented on src/pocketmine/network/mcpe/protocol/StartGamePacket.php:
```diff
@@ -130,6 +134,8 @@ class StartGamePacket extends DataPacket{
 	public $onlySpawnV1Villagers = false;
 
+	/** @var string */
+	public $vanillaVersion = ProtocolInfo::MINECRAFT_VERSION_NETWORK;
```
I think it’s better to leave an empty string here - the buffer will be a couple of bytes less. This field is useless.

Also `vanillaVersion` refers to LevelSettings, you put newline in the wrong place.

SupremeMortal replied:
From what I've heard, sending anything other than the version will crash the client.

yoraze replied:
Personally, I send an empty string there and everything works fine...

dktapps replied:
works for you != works for everyone

haggling over a premature optimisation like that is a waste of everyone's time.

yoraze replied:
> works for you != works for everyone

But what if I say that everyone who uses my fork with 1.13 support works fine?
By the way, I want to note that BDS itself sends an empty string there.

dktapps replied:
It doesn't matter either way. This is just bikeshedding over a micro optimisation. Let it go.

yoraze commented on src/pocketmine/network/mcpe/NetworkBinaryStream.php:
```diff
@@ -74,6 +77,62 @@ public function putUUID(UUID $uuid) : void{
+		$this->putString($skin->getFullSkinId());
+	}
+
+	public function getImage() : SerializedImage{
```
But this does not exist in the protocol...
If you have a PC and can download the IDA, take a look at the serialize method in SerializedSkin.

dktapps replied:
This comment is nonsense. The presence and location of this function have absolutely zero bearing on the protocol.

yoraze commented on src/pocketmine/entity/Skin.php:
```diff
@@ -120,6 +152,48 @@ public function getGeometryData() : string{
+	 * @return string
+	 */
+	public function getFullSkinId() : string{
+		return $this->skinId . "_" . $this->capeId;
```
You could leave an empty string here, it is useless anyway.
However, I want to note that the client sends SkinId here again.

Sandertv replied:
The cape ID sent is (currently) always empty.

Commit:
Empty fields in PlayerSkinPacket, thanks @Minejong

Commit:
Use unknown os for PlayerListEntry

Commit:
reduce visibility of offset to private

dktapps commented on src/pocketmine/network/mcpe/protocol/PlayerSkinPacket.php:
```diff
@@ -38,15 +38,23 @@ class PlayerSkinPacket extends DataPacket{
+	/** @var string */
+	public $newSkinName = "";
+	/** @var string */
+	public $unknownString = ""; //Sent as empty, assuming it is the old skin name
```
This should be `oldSkinName` as per the original code.

dktapps commented on src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter.php:
```diff
@@ -248,7 +253,7 @@ public function handleResourcePackChunkRequest(ResourcePackChunkRequestPacket $p
 
 	public function handlePlayerSkin(PlayerSkinPacket $packet) : bool{
-		return $this->player->changeSkin($packet->skin, $packet->newSkinName, $packet->oldSkinName);
+		return $this->player->changeSkin($packet->skin, "", "");
```
This change should be removed.

dktapps committed:
Merge remote-tracking branch 'origin/stable' into drew-1.13

dktapps committed:
auto generated baseline

Commit:
Update playerskinpacket to use the strings

Commit:
Friendly BC skins (persona not supported)

Wumpotamus commented:
As of the latest commit, legacy skins are the only working skins, persona skins are invisible and there's a weird bug that if you pause the screen, your skin goes invisible. As discussed on discord, the plan is to convert persona skins to a legacy skin, of a random color

Commit:
Remove unused imports and unused todo

Commit:
Update SkinAnimation constants and docs

Commit:
change buildPlatform parameter to -1

Commit:
Clean up unused imports

dktapps commented on src/pocketmine/network/mcpe/NetworkBinaryStream.php:
```diff
@@ -74,6 +77,62 @@ public function putUUID(UUID $uuid) : void{
+
+	public function putSkin(Skin $skin){
+		$this->putString($skin->getSkinId());
+		$this->putString($skin->getGeometryName()); //resource patch
```
doesn't this need converting to a skin resource patch?

dktapps commented on src/pocketmine/utils/SerializedImage.php:
```diff
@@ -0,0 +1,67 @@
+
+declare(strict_types=1);
+
+namespace pocketmine\utils;
```
This should be moved to the `pocketmine\network\mcpe\protocol\types` namespace.

dktapps commented on src/pocketmine/utils/SkinAnimation.php:
```diff
@@ -0,0 +1,71 @@
+
+declare(strict_types=1);
+
+namespace pocketmine\utils;
```
This should be moved to the `pocketmine\network\mcpe\protocol\types` namespace.

dktapps reviewed (changes requested):
My recommendation was to provide a separate protocol-specific `Skin` type, which would be converted to a legacy skin at packet handle and before sending.

We don't want to lose information from packets when debugging. The current BC impl simply discards information it doesn't find useful.

dktapps commented on src/pocketmine/utils/SerializedImage.php:
```diff
@@ -0,0 +1,67 @@
+
+namespace pocketmine\utils;
+
+class SerializedImage{
```
This probably should be named `SkinImage`.

dktapps committed:
PlayerSkinPacket: reduce diff pollution

dktapps commented on src/pocketmine/network/mcpe/NetworkBinaryStream.php:
```diff
@@ -74,6 +77,62 @@ public function putUUID(UUID $uuid) : void{
+			$this->putLInt($animation->getType());
+			$this->putLFloat($animation->getFrames());
+		}
+		$this->putImage(new SerializedImage(0, 0, $skin->getCapeData()));
```
This looks likely to cause problems for non-null capes.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -2913,6 +2911,14 @@ public function handleAnimate(AnimatePacket $packet) : bool{
+			$this->sendRespawnPacket($this, RespawnPacket::READY_TO_SPAWN);
+		}
+
+		return true;
```
This should return unhandled if it arrived in an unexpected condition.

dktapps commented on src/pocketmine/entity/Skin.php:
```diff
@@ -55,6 +55,10 @@ public function __construct(string $skinId, string $skinData, string $capeData =
 	}
 
+	public static function convertToLegacyName(string $name) : string{
+		return '{"geometry" : {"default" : "' . $name . '"}}';
```
This looks susceptible to injection bugs.
```suggestion
		return json_encode(["geometry" => ["default" => $name]]);
```

dktapps commented on src/pocketmine/level/particle/FloatingTextParticle.php:
```diff
@@ -96,7 +96,7 @@ public function encode(){
 			$add = new PlayerListPacket();
 			$add->type = PlayerListPacket::TYPE_ADD;
-			$add->entries = [PlayerListEntry::createAdditionEntry($uuid, $this->entityId, $name, new Skin("Standard_Custom", str_repeat("\x00", 8192)))];
+			$add->entries = [PlayerListEntry::createAdditionEntry($uuid, $this->entityId, $name, new Skin("Standard_Custom", str_repeat("\x00", 8192), "", Skin::convertToLegacyName("geometry.humanoid.custom"), ""))];
```
Is this change still necessary with BC core skins?

dktapps committed:
SkinAnimation: fix typo

dktapps commented on src/pocketmine/network/mcpe/NetworkBinaryStream.php:
```diff
@@ -74,6 +77,62 @@ public function putUUID(UUID $uuid) : void{
+		$animationCount = $this->getLInt();
+		$animations = [];
+		for($i = 0; $i < $animationCount; ++$i){
+			$animations[] = new SkinAnimation($this->getImage(), $this->getLInt(), $this->getLFloat());
```
For the sake of code clarity, please make the following change:
```suggestion
			$animations[] = new SkinAnimation(
				$image = $this->getImage(),
				$animationType = $this->getLInt(),
				$animationFrames = $this->getLFloat()
			);
```
This makes it clear what the purpose of the fields being read are, and refactoring will persist these assignments.

dktapps commented on src/pocketmine/network/mcpe/NetworkBinaryStream.php:
```diff
@@ -74,6 +77,62 @@ public function putUUID(UUID $uuid) : void{
+		$this->putString(""); //fullskinId
+	}
+
+	public function getImage() : SerializedImage{
```
This should be renamed to `getSkinImage()` to avoid confusion.
It's also not clear to me that this needed to be public.

dktapps commented on src/pocketmine/network/mcpe/NetworkBinaryStream.php:
```diff
@@ -74,6 +77,62 @@ public function putUUID(UUID $uuid) : void{
+		return new SerializedImage($height, $width, $data);
+	}
+
+	public function putImage(SerializedImage $image) : void{
```
This should be renamed to `putSkinImage()` to avoid confusion.
It's also not clear to me that this needed to be public.

dktapps committed:
Merge branch 'stable' of https://github.com/drew-mcbe/pocketmine-mp into drew-1.13

dktapps committed:
Merge remote-tracking branch 'origin/stable' into drew-1.13

dktapps committed:
add encode/decode for some new packets

dktapps committed:
ResourcePackStack: missing field

Commit:
Merge pull request #4 from dktapps/drew-1.13

moar 1.13 changes

Commit:
Move SkinAnimation and changed SerializedImage to SkinImage

Also change putImage and getImage to getSkinImage and putSkinImage

Commit:
Remove changes to FloatingTextParticle

Commit:
Remove convertToLegacyName

Commit:
Change get and put SkinImage visibility to private

Commit:
Added variables for getSkin's animation for easier readability

dktapps commented on src/pocketmine/network/mcpe/NetworkBinaryStream.php:
```diff
@@ -74,6 +77,66 @@ public function putUUID(UUID $uuid) : void{
+		$capeId = $this->getString();
+		$fullSkinId = $this->getString();
+
+		return new Skin($skinId, $skinData->getData(), $capeData->getData(), $skinResourcePatch, $geometryData);
```
the geometry name should be extracted from the resource patch `default` field, as far as I understand.

Commit:
Return false in unexpected condition

dktapps committed:
AddPlayerPacket: missing field

dktapps committed:
Update Player.php

dktapps committed:
Merge branch 'stable' of https://github.com/drew-mcbe/pocketmine-mp into drew-1.13

dktapps committed:
fix skin sharing stupidity

dktapps committed:
MultiplayerSettingsPacket

dktapps committed:
fix classic capes

Commit:
Merge pull request #5 from dktapps/drew-1.13

more 1.13

Commit:
Added network components for skins, to collect instead of throw out data

Commit:
Move persona skin hack to legacyskinadapter

Commit:
Extract geometry name from resource patch

Commit:
Update AvailableCommandsPacket constants

dktapps commented:
Latest commits are starting to look a little more interesting, but the geometry name handling is still incorrect. It should be extracted from `skinResourcePatch["geometry"]["default"]`.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/LegacySkinAdapter.php:
```diff
@@ -32,6 +32,10 @@ public function toSkinData(Skin $skin) : SkinData{
+		if($data->persona){
+			$skinData = str_repeat(random_bytes(3) . "\xff", 2048);
+		}
+		return new Skin($data->skinId, $skinData, $data->capeImage->getData(), $data->resourcePatch, $data->geometryData);
```
There may need to be additional checks for `personaCapeOnClassicSkin`.

dktapps commented on src/pocketmine/network/mcpe/protocol/PlayerListPacket.php:
```diff
@@ -89,13 +79,12 @@ protected function encodePayload(){
 				$this->putString($entry->xboxUserId);
 				$this->putString($entry->platformChatId);
+				$this->putLInt($entry->buildPlatform);
+				$this->putSkin($entry->skinData);
```
this looks like a bug

dktapps commented on src/pocketmine/network/mcpe/protocol/PlayerListPacket.php:
```diff
@@ -57,22 +57,12 @@ protected function decodePayload(){
 				$entry->xboxUserId = $this->getString();
 				$entry->platformChatId = $this->getString();
+				$entry->buildPlatform = $this->getLInt();
+				$entry->skin = $this->getSkin();
```
this looks like a bug

dktapps commented on src/pocketmine/network/mcpe/protocol/types/LegacySkinAdapter.php:
```diff
@@ -0,0 +1,41 @@
+		if($data->persona){
+			$skinData = str_repeat(random_bytes(3) . "\xff", 2048);
+		}
+		return new Skin($data->skinId, $skinData, $data->capeImage->getData(), json_decode($data->resourcePatch, true)["geometry"]["default"], $data->geometryData);
```
I wouldn't assume the `resourcePatch` is valid; it might result in an uncaught error and crash.

dktapps replied:
It's possible to have a Persona cape attached to a classic skin. Capes should be removed if they aren't classic.

dktapps commented on src/pocketmine/level/particle/FloatingTextParticle.php:
```diff
@@ -96,7 +97,9 @@ public function encode(){
 			$add->type = PlayerListPacket::TYPE_ADD;
-			$add->entries = [PlayerListEntry::createAdditionEntry($uuid, $this->entityId, $name, new Skin("Standard_Custom", str_repeat("\x00", 8192)))];
+			$add->entries = [PlayerListEntry::createAdditionEntry($uuid, $this->entityId, $name, SkinAdapterSingleton::get()->toSkinData(new Skin(
+				"Standard_Custom", str_repeat("\x00", 8192), "", '{"geometry" : {"default" : "geometry.humanoid.custom"}}' //TODO: Remove, hack to fix ftp
```
It's not clear to me what the purpose of this hack is, and to my understanding of the code, this hardcoded JSON should not be here.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/SkinAdapterSingleton.php:
```diff
@@ -0,0 +1,40 @@
+
+class SkinAdapterSingleton{
+
+	private static $skinAdapter = null;
```
Missing `@var` doc comment here.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/LegacySkinAdapter.php:
```diff
@@ -0,0 +1,41 @@
+	public function fromSkinData(SkinData $data) : Skin{
+		$skinData = $data->skinImage->getData();
+		if($data->persona){
+			$skinData = str_repeat(random_bytes(3) . "\xff", 2048);
```
This is not sufficient. The geometry model information, persona cape and more will persist, which is likely the reason for your headless players.
```suggestion
			return new Skin("Standard_Custom", str_repeat(random_bytes(3) . "\xff", 2048), "", "geometry.humanoid.custom");
```

dktapps commented on src/pocketmine/network/mcpe/protocol/types/SkinData.php:
```diff
@@ -0,0 +1,64 @@
+	/** @var bool */
+	public $premium;
+	/** @var bool */
+	public $capeOnClassic;
```
This would be better named `personaCapeOnClassicSkin`.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/SkinData.php:
```diff
@@ -0,0 +1,64 @@
+	/** @var bool */
+	public $capeOnClassic;
+	/** @var string */
+	public $capeId;
```
These fields shouldn't be public; please make them private and generate getters for them.

dktapps commented on src/pocketmine/network/mcpe/protocol/types/SkinData.php:
```diff
@@ -0,0 +1,64 @@
+	/** @var string */
+	public $capeId;
+
+	public function __construct(string $skinId, string $resourcePatch, SkinImage $skinImage, array $animations = [], SkinImage $capeImage = null, string $geometryData = "", string $animationData = "",  bool $premium = false, bool $persona = false, bool $capeOnClassic = false, string $capeId = ""){
```
Please provide `@param` docs here; there is an `array` parameter whose type isn't clear from the native typehints.

dktapps replied:
Again here: `capeOnClassic` would be better named `personaCapeOnClassicSkin`.

dktapps commented:
There are some additional protocol changes which have not been addressed, which can be seen in this diff: https://github.com/Sandertv/gophertunnel/commit/139aa27d2bc8ea4755b6cf2e4ace71c647aab461

- `CraftingDataPacket` has two additional varint-prefixed lists before the end of the packet containing `PotionMix` and `PotionContainerMix` structures (paraphrasing, pls choose better names if you can think of any):
  - `PotionRecipe` (for brewing) has an input potion ID (item metadata), ingredient item ID, and output potion ID (item metadata)
  - `PotionContainerMix` (for e.g. converting a normal potion to splash potion) has an input item ID, ingredient item ID, and output item ID
- There is an extra list of shit at the end of `AvailableCommandsPacket` that doesn't make much sense right now. At the least, it should be filled out with a zero length and a TODO placed.
- `isEduMode` in `StartGamePacket` appears to have changed; see https://github.com/Sandertv/gophertunnel/commit/139aa27d2bc8ea4755b6cf2e4ace71c647aab461#diff-daacd64fd2e506c3704bcabd09f4636aR65

Commit:
Generate getters for SkinData, applied suggested change, and fixed an underfined variable in PlayerListPacket

Commit:
Fixed floatingtextparticle yet again

Commit:
make toSkinData return statement multi lined

Commit:
Remove persona capes off of classic skins, add checks for empty cape data when converting

Commit:
Should be checking the string, not an object

Commit:
Generate param docs for constructor

Commit:
Remove inline assignment

Commit:
Changed eduMode to eduEditionOffer and added an unsignedVarInt in availablecommandspacket

yoraze commented on src/pocketmine/network/mcpe/protocol/StartGamePacket.php:
```diff
@@ -83,8 +83,8 @@ class StartGamePacket extends DataPacket{
-	/** @var bool */
-	public $eduMode = false;
+	/** @var int */
+	public $eduEditionOffer = 0;
```
I ask you to make some changes to Player about this. (In Player you can find sending StartGamePacket and do something with it, otherwise there will be extra crashdumps :P)

Wumpotamus replied:
Thanks

Commit:
Added : void typehint to SkinAdapterSingleton & provide documentation

Commit:
Reflect StartGamePacket changes in Player

Commit:
Add some isset checks to prevent server from crashing on an invalid resourcepatch

dktapps commented on src/pocketmine/network/mcpe/protocol/types/LegacySkinAdapter.php:
```diff
@@ -43,11 +43,16 @@ public function toSkinData(Skin $skin) : SkinData{
 		$capeData = $data->getCapeImage()->getData();
+		$geometryName = "";
+		$resourcePatch = json_decode($data->getResourcePatch(), true);
+		if(isset($resourcePatch["geometry"]) && isset($resourcePatch["geometry"]["default"])){
```
This isn't sufficient. The `geometry` key might not be an array, or the `default` key might not be a string.

dktapps replied:
In addition, the player should be rejected with `Invalid skin` if the resource patch is invalid.

Commit:
Remove offset for crafting grid

Commit:
Better checks on resource patch

Wumpotamus commented:
You don't, this isn't ready for production and shouldn't be compiled

Wumpotamus commented:
note for later:

Inventory offsets go as:
```
Cursor - 0
Anvil - 1
Enchant - 14
Beacon - 27
Crafting Small - 28
Crafting Big - 32
```

dktapps commented:
seems like there are some gaps?

nituzhazha commented:
I used the method of setSkin() to set the skin, but the skin could not be shown in the test. Is this the core problem or I may have written it wrong? Please tell me

dktapps commented:
This isn't a plugin support forum.

nituzhazha commented:
> This isn't a plugin support forum.

I'm sorry...

dktapps committed:
Merge branch 'stable' into drew-1.13

dktapps committed:
AvailableCommandsPacket: encode & decode for enum value constraints
This is a peculiarly overengineered system that is used for restricting access to enum members under certain conditions, e.g. to disallow changing specific gamerules in survival.

dktapps committed:
AvailableCommandsPacket: add special handling for enums which aren't referenced by any command directly
the CommandName enum is a magic enum used by the  argtype.

TODO: It's possible that not sending the CommandName enum might be causing client sided crashes. Investigate.

dktapps committed:
MoveActorDeltaPacket: flags is now a short

dktapps committed:
Merge branch 'stable' into drew-1.13

dktapps committed:
CraftingDataPacket: read & write potion recipes

Commit:
Merge remote-tracking branch 'upstream/stable' into stable

Commit:
Merge branch 'drew-1.13' of https://github.com/dktapps/PocketMine-MP into dktapps-drew-1.13

Commit:
Merge branch 'dktapps-drew-1.13' into stable

dktapps committed:
AvailableCommandsPacket: fixed doc comment for putEnumConstraint()

dktapps committed:
RuntimeBlockMapping: add a type check for decoded NBT root type
this is kinda redundant since this function can blow up in so many other ways anyway, but it makes PHPStan happy, so it'll do.

dktapps committed:
AvailableCommandsPacket: fix nonsensical placement of HARDCODED_ENUMS

dktapps committed:
EducationSettingsPacket: added encode & decode

dktapps committed:
PlayerAuthInputPacket: added encode & decode, and some aux classes

dktapps committed:
StartGamePacket: fixed asymmetry in block table handling

dktapps commented:
Doors and item frame placement is broken; doors don't hinge correctly and item frames are ... screwed.

dktapps committed:
inventory: fix some transactions being rejected for no good reason
since 1.13, transactions such as interacting with creative inventory cause a spoof windowID 124 slot 50 action to appear which changes air -> air. This currently gets rejected because only cursor is mapped to ID 124, and it only has a single slot.

It is not clear what the purpose of 124:50 is, but this fix filters out any actions which do not change anything, since they won't affect transaction balance anyway.

dktapps committed:
crafting now works on 1.13, but it's not good

dktapps commented:
There has been further discussion on the skins issue, and it's come to light that the `SkinAdapter` may not be sufficient for third party developers to bolt on Persona support properly. In addition, players think that the solid colour skins are a bug.

Given that there is no way to implement Persona properly without a clean BC break, it may be necessary to release a new major version to support 1.13, which has skin changes and nothing else.

dktapps committed:
Merge branch 'stable' into drew-1.13

dktapps committed:
StartGamePacket: fixed possible type violation on decoding block table

dktapps committed:
updated submodule to pmmp/BedrockData@a38b42788883fa2094f67874f15594044be1ac4d

dktapps committed:
RuntimeBlockMapping: use new data, item frames & doors fixed
floor & ceiling item frames not supported though

dktapps committed:
AvailableEntityIdentifiersPacket: load pregenerated data from file

dktapps committed:
BiomeDefinitionListPacket: load pregenerated data from file

dktapps committed:
InventoryTransactionPacket: fixed crafting flags being set on useless empty transactions
this fixes some misleading debug noise when getting stuff from creative inventory and triggers some different spam instead.

superbobby2000 commented:
I don't understand how PlayerListEntry work now
